File: js/board.js
class Board {
  constructor(size) {
    this.size = size;
    this.end = false;
    this.map = this.createMap();
    this.state = [];
    for (let i = 0; i < (size * size); i++) {
      this.state.push(0);
    }
  }

  initiateState(state) {
    this.state = JSON.parse(JSON.stringify(state));
  }

  input(pos, ID) {
    if (pos >= 0 && pos < (this.size * this.size)) {
      if (this.state[pos] == 0) {
        this.state[pos] = ID;
        return this.state;
      } else {
        console.log("ID: " + ID + "   " + pos + " is already taken.");
        return NaN;
      }
    } else if (pos == -1) {
      console.log("Passed");
      return this.state;
    } else {
      console.log("invalid position");
      return NaN;
    }
  }

  drawBoard() {
    var rowElement;
    //goes through y axis
    for (var y = 0; y < size; y++) {
      rowElement = $("<div class='board-row'></div>");
      //goes through x axis
      for (var x = 0; x < size; x++) {
        var cellElement = $("<div type='button' id='" + x + "-" + y + "' class='cells unselected'><img src='images/gameAssets/unselected.png'></div>");
        rowElement = rowElement.append(cellElement);
      }
      $("#gameboard").append(rowElement);
    }
  }

  isEnd() {
    let max = 0;
    let leadID = 0;
    for (let i = 0; i < this.size * this.size; i++) {
      let targetID = this.state[i];
      if (targetID == 0) {
        continue;
      }

      for (let d = 0; d < 8; d++) {
        let counter = 1;
        let currentPos = i;
        while (targetID == this.getID(this.getNextPos(currentPos, d))) {
          counter += 1;
          currentPos = this.getNextPos(currentPos, d);
        }

        if (counter > max) {
          max = counter;
          leadID = targetID;
          if (counter >= this.size) {
            this.end = true;
            return leadID;
          }
        }
      }
    }
    return leadID;
  }


  getID(pos) {
    if (pos >= 0 && pos < (this.size * this.size)) {
      return this.state[pos];
    }
    return 0;
  }


  getNextPos(pos, direction) {
    let toReturn = (this.map[pos][direction] != NaN) ? (this.map[pos][direction] + pos) : -1;
    return toReturn;
  }



  createMap() {
    //directions are 8 directions (0-7) from up clockwise.
    var index = 0;
    var mapping = [];
    for (let i = 0; i < this.size * this.size; i++) {
      mapping.push([]);
    }
    for (var y = 0; y < size; y++) {
      for (var x = 0; x < size; x++) {
        //bottom edge
        if ((y + 1) % size == 0) {
          //right edge
          if ((x + 1) % size == 0) {
            mapping[index] = [-size, NaN, NaN, NaN, NaN, NaN, -1, -(size + 1)]

            //left edge
          } else if (x == 0) {
            mapping[index] = [-size, -(size - 1), +1, NaN, NaN, NaN, NaN, NaN]

            //other safe places
          } else {
            mapping[index] = [-size, -(size - 1), +1, NaN, NaN, NaN, -1, -(size + 1)]
          }

          //top edge
        } else if (y == 0) {
          //right edge
          if ((x + 1) % size == 0) {
            mapping[index] = [NaN, NaN, NaN, NaN, size, size - 1, -1, NaN]

            //left edge
          } else if (x == 0) {
            mapping[index] = [NaN, NaN, +1, size + 1, size, NaN, NaN, NaN]

            //other safe places
          } else {
            mapping[index] = [NaN, NaN, +1, size + 1, size, size - 1, -1, NaN]
          }

          //other safe places
        } else {
          //right edge
          if ((x + 1) % size == 0) {
            mapping[index] = [-size, NaN, NaN, NaN, size, size - 1, -1, -(size + 1)]

            //left edge
          } else if (x == 0) {
            mapping[index] = [-size, -(size - 1), +1, size + 1, size, NaN, NaN, NaN]

            //other safe places
          } else {
            mapping[index] = [-size, -(size - 1), +1, size + 1, size, size - 1, -1, -(size + 1)]
          }
        }
        index++;
      }
    }
    return mapping;
  }

}

File: python/board.py
import copy

class Board:
  def __init__(self, size,connect):
    self.size = size
    self.connect = connect
    self.end = False
    self.map = self.createMap()
    self.state = []
    for i in range(size * size):
      self.state.append(0)
  
  def initiateState(self,state):
    self.state = copy.deepcopy(state)
  
  def input(self,pos,ID):
    if pos >= 0 and pos < (self.size * self.size):
      if(self.state[pos] == 0):
        self.state[pos] = ID
        return self.state
      else:
        print("ID: " + str(ID) +"   "+ str(pos) + " is already taken.")
        return None
    elif pos == -1:
      print("Passed")
      return self.state
    else:
      print("invalid position")
      return None

  def drawBoard(self):
    toPrint = ""
    for i in range(self.size):
      toPrint += "   " + str(i + 1)
    toPrint += "\n"
    for i in range(self.size):
      toPrint += "-----"
    for y in range(self.size):
      toPrint += "\n"
      toPrint += str(y + 1) + "|  "
      for x in range(self.size):
        target = self.getID(y * self.size + x)
        drawing = "-"
        if target == -1:
          drawing = "○"
        elif target == 1:
          drawing = "⬤"
        toPrint += drawing + " | "
    print(toPrint)
  
  def isEnd(self):
    max = 0
    leadID = 0
    

    for i in range(self.size*self.size):

      targetID = self.state[i]

      #not owned by anyone
      if targetID == 0:
        continue

      #0-7 direction (up, upright, right ....)
      for d in range(8):
        counter = 1
        currentPos = i
        while (targetID == self.getID(self.getNextPos(currentPos,d))):
          counter += 1
          currentPos = self.getNextPos(currentPos,d)
        
        if counter > max:
          max = counter
          leadID = targetID

          if counter >= self.connect:
            self.end = True
            return leadID
    return leadID

  def getID(self,pos):
    if pos >= 0 and pos < (self.size * self.size):
      return self.state[pos]
    return 0
  
  def getNextPos(self,pos,direction):
    return (self.map[pos][direction] + pos) if (not self.map[pos][direction] == None) else -1


  def createMap(self):
    index = 0
    size = self.size
    mapping = []
    for n in range(size*size):
      mapping.append([])
    for y in range(size):
      for x in range(size):
        #bottom edge
        if((y + 1) % size == 0):
          #right edge
          if((x + 1) % size == 0):
            mapping[index] = [-size, None, None, None, None, None, -1, -(size+1)]

            #left edge
          elif(x == 0):
            mapping[index] = [-size,-(size-1),+1,None,None,None,None,None ]

            #other safe places
          else:
            mapping[index] = [-size,-(size-1),+1,None,None,None,-1,-(size+1)]

          #top edge
        elif(y == 0):
          #right edge
          if((x + 1) % size == 0):
            mapping[index] = [None,None,None,None,size,size-1,-1,None]

            #left edge
          elif(x == 0):
            mapping[index] = [None,None,+1,size+1,size,None,None,None]

            #other safe places
          else:
            mapping[index] = [None,None,+1,size+1,size,size-1,-1,None]

          #other safe places
        else:
          #right edge
          if((x + 1) % size == 0):
            mapping[index] = [-size,None,None,None,size,size-1,-1,-(size+1)]

            #left edge
          elif(x == 0):
            mapping[index] = [-size,-(size-1),+1,size+1,size,None,None,None]

            #other safe places
          else:
            mapping[index] = [-size,-(size-1),+1,size+1,size,size-1,-1,-(size+1)]

        index += 1
    return mapping

  

